pkg/pub_package_reader/lib/pub_package_reader.dart

```diff
@@ -310,6 +311,44 @@ Iterable<ArchiveIssue> checkAuthors(String pubspecContent) sync* {
   }
 }
 
+final _strictRegExp = RegExp(r'^' // Start at beginning.
+    r'(\d+)\.(\d+)\.(\d+)' // Version number.
+    r'(-([0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*))?' // Pre-release.
+    r'(\+([0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*))?' // Build.
+    r'$' // End of string.
+    );
+
+/// Checks if all version constraints follow a stricter pattern.
+Iterable<ArchiveIssue> checkStrictVersions(Pubspec pubspec) sync* {
+  Iterable<Version?> expandConstraint(VersionConstraint? constraint) {
+    if (constraint is VersionRange) {
+      return [constraint.min, constraint.max];
+    }
+    return const Iterable.empty();
+  }
+
+  Iterable<Version?> expandDependency(Dependency? dependency) {
+    if (dependency is HostedDependency) {
```

Comment:
All dependency types can have `version` constraints... it's just a bug in `pubspec_parse` that it can't understand this...

This is a probably a temporary fix anyways, once we upgrade `pub_semver` we won't need this hack anymore.

(filed https://github.com/dart-lang/pubspec_parse/issues/77)

Reply:
Added a TODO for it.